In [1]:
import pandas as pd

In [2]:
foncier2014 = pd.read_csv('../../data/foncier/full_2014.csv',sep=',',dtype= {'code_commune' : str})

/home/yopla/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (12,14,16,17,18,20,22,35,36) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
foncier2014.head()

id_mutation date_mutation  numero_disposition nature_mutation  \
0      2014-1    2014-01-09                   1           Vente   
1      2014-2    2014-01-09                   1           Vente   
2      2014-2    2014-01-09                   1           Vente   
3      2014-3    2014-01-02                   1           Vente   
4      2014-3    2014-01-02                   1           Vente   

   valeur_fonciere  adresse_numero adresse_suffixe   adresse_nom_voie  \
0         251500.0           532.0             NaN    RTE DE MEZERIAT   
1         174500.0            22.0             NaN     BD MAL LECLERC   
2         174500.0            22.0             NaN     BD MAL LECLERC   
3         157500.0          1389.0             NaN  RTE DE STRASBOURG   
4         157500.0          1369.0             NaN  RTE DE STRASBOURG   

  adresse_code_voie  code_postal  ...   type_local surface_reelle_bati  \
0              0006       1310.0  ...       Maison               147.0   
1              2540       1000.0  ...  Appartement                80.0   
2              2540       1000.0  ...   Dépendance                 NaN   
3              0201       1440.0  ...          NaN                 NaN   
4              0201       1440.0  ...       Maison               103.0   

  nombre_pieces_principales  code_nature_culture nature_culture  \
0                       5.0                    S           sols   
1                       2.0                  NaN            NaN   
2                       0.0                  NaN            NaN   
3                       NaN                    S           sols   
4                       4.0                    S           sols   

  code_nature_culture_speciale nature_culture_speciale surface_terrain  \
0                          NaN                     NaN          1501.0   
1                          NaN                     NaN             NaN   
2                          NaN                     NaN             NaN   
3                          NaN                     NaN            24.0   
4                          NaN                     NaN          1569.0   

  longitude   latitude  
0  5.119191  46.247576  
1  5.218208  46.204017  
2  5.218208  46.204017  
3  5.262505  46.236179  
4  5.262756  46.236276  

[5 rows x 40 columns]

In [4]:
foncier2014['valeur_fonciere'].isna().sum()

28279

In [5]:
res = foncier2014['type_local'].unique()
res

array(['Maison', 'Appartement', 'Dépendance', nan,
       'Local industriel. commercial ou assimilé'], dtype=object)

In [6]:
foncier2014.type_local.value_counts(dropna=False)

NaN                                         1169548
Maison                                       470502
Appartement                                  441343
Dépendance                                   344057
Local industriel. commercial ou assimilé      91238
Name: type_local, dtype: int64

In [7]:
immo2014 = foncier2014

# pour faire un premier tri dans les colonnes

In [8]:
cols = ['id_mutation', 'date_mutation', 'numero_disposition',
       'nature_mutation', 'valeur_fonciere', 
       'code_postal', 'code_commune', 'nom_commune', 'code_departement', 'id_parcelle',
        'code_type_local', 'type_local',
       'surface_reelle_bati', 'nombre_pieces_principales',
       'code_nature_culture', 'nature_culture',
    'surface_terrain', 'longitude', 'latitude']

In [9]:
immo2014 = immo2014[cols]

# pour faire un premier dans les lignes

### On ne garde pas les ligne où type local est nulle

In [12]:
immo2014 = immo2014[immo2014.type_local.notnull()]

### On dégage les  'Dépendances'

Il y a 1040 dépendances avec une surface réelle batie donnée contre 343017 ( 99.69 % de NA ).
On ne va pas prendre en comptes les dépendances. 

In [14]:
immo2014[immo2014.code_type_local == 3 ]["surface_reelle_bati"].isna().value_counts(dropna=False)

True     343017
False      1040
Name: surface_reelle_bati, dtype: int64

In [15]:
immo2014 = immo2014[immo2014.code_type_local != 3 ]

### On dégage les  'Local industriel. commercial ou assimilé'

In [16]:
immo2014[immo2014.code_type_local == 4 ]["surface_reelle_bati"].isna().value_counts(dropna=False)

False    68978
True     22260
Name: surface_reelle_bati, dtype: int64

In [17]:
immo2014 = immo2014[immo2014.code_type_local != 4 ]

Il y a 22260 'Local industriel. commercial ou assimilé' avec une surface réelle batie donnée contre 68978 ( 75 % de NA ). 
On ne va pas prendre en comptes les 'Local industriel. commercial ou assimilé'. 

### On dégage les  biens avec une valeur foncière pas donnée 

In [18]:
immo2014.valeur_fonciere.isna().value_counts()

False    906100
True       5745
Name: valeur_fonciere, dtype: int64

Il y a 0.63 de Na ( 5745 sur 906100 ) sur la valeur foncière pour les appartements et les maisons.
On enlève les lignes où la valeur foncière est nulle dans un premier temps

In [19]:
immo2014 = immo2014[immo2014.valeur_fonciere.notnull()]

### On dégage les biens avec une surface_reelle_bati = NA et un nombre de pices principales  = NA
cela concerne 64 ligne sur 906000 seulement

In [20]:
immo2014.isna().sum()

id_mutation                       0
date_mutation                     0
numero_disposition                0
nature_mutation                   0
valeur_fonciere                   0
code_postal                    1851
code_commune                      0
nom_commune                       0
code_departement                  0
id_parcelle                       0
code_type_local                   0
type_local                        0
surface_reelle_bati              65
nombre_pieces_principales        64
code_nature_culture          377667
nature_culture               377667
surface_terrain              377675
longitude                     36756
latitude                      36756
dtype: int64

In [21]:
immo2014 = immo2014[immo2014.surface_reelle_bati.notnull()]

In [22]:
immo2014.isna().sum()

id_mutation                       0
date_mutation                     0
numero_disposition                0
nature_mutation                   0
valeur_fonciere                   0
code_postal                    1851
code_commune                      0
nom_commune                       0
code_departement                  0
id_parcelle                       0
code_type_local                   0
type_local                        0
surface_reelle_bati               0
nombre_pieces_principales         0
code_nature_culture          377611
nature_culture               377611
surface_terrain              377619
longitude                     36751
latitude                      36751
dtype: int64

## On enlève les duplicates

####  avec une règle maison ( on prend le premier élement des groupes fait sur l'id de mutation )

In [25]:
nb_id_mutation = immo2014.groupby(['id_mutation']).size()
nonunique_id_mutation = nb_id_mutation[nb_id_mutation != 1]
nonunique_id_mutation = nonunique_id_mutation.index

In [26]:
immo2014[immo2014.id_mutation.isin(nonunique_id_mutation)].sort_values(by=['id_mutation'], ascending=False)

id_mutation date_mutation  numero_disposition  \
2358982   2014-999991    2014-10-27                   1   
2358981   2014-999991    2014-10-27                   1   
2358965   2014-999980    2014-10-30                   1   
2358963   2014-999980    2014-10-30                   1   
2358955   2014-999980    2014-10-30                   1   
2358956   2014-999980    2014-10-30                   1   
2358957   2014-999980    2014-10-30                   1   
2358964   2014-999980    2014-10-30                   1   
248570     2014-99998    2014-05-16                   1   
248569     2014-99998    2014-05-16                   1   
2358918   2014-999967    2014-10-24                   1   
2358919   2014-999967    2014-10-24                   1   
2358861   2014-999936    2014-10-31                   1   
2358859   2014-999936    2014-10-31                   1   
2358851   2014-999931    2014-10-31                   1   
2358850   2014-999931    2014-10-31                   1   
2358778   2014-999882    2014-11-12                   1   
2358777   2014-999882    2014-11-12                   1   
2358776   2014-999882    2014-11-12                   1   
2358773   2014-999881    2014-04-08                   1   
2358774   2014-999881    2014-04-08                   1   
2358748   2014-999864    2014-10-16                   1   
2358747   2014-999864    2014-10-16                   1   
2358743   2014-999860    2014-10-27                   1   
2358742   2014-999860    2014-10-27                   1   
2358682   2014-999823    2014-10-31                   1   
2358681   2014-999823    2014-10-31                   1   
2358652   2014-999804    2014-10-23                   1   
2358647   2014-999804    2014-10-23                   1   
2358646   2014-999804    2014-10-23                   1   
...               ...           ...                 ...   
2359289  2014-1000146    2014-11-06                   1   
2359288  2014-1000146    2014-11-06                   1   
2359287  2014-1000146    2014-11-06                   1   
2359286  2014-1000146    2014-11-06                   1   
2359234  2014-1000125    2014-11-14                   1   
2359233  2014-1000125    2014-11-14                   1   
2359232  2014-1000125    2014-11-14                   1   
2359188  2014-1000102    2014-11-07                   1   
2359189  2014-1000102    2014-11-07                   1   
2359165  2014-1000092    2014-11-05                   2   
2359166  2014-1000092    2014-11-05                   2   
2359164  2014-1000092    2014-11-05                   2   
2359163  2014-1000092    2014-11-05                   2   
2359162  2014-1000092    2014-11-05                   2   
2359161  2014-1000092    2014-11-05                   2   
2359159  2014-1000092    2014-11-05                   2   
2359158  2014-1000092    2014-11-05                   2   
2359081  2014-1000042    2014-10-30                   1   
2359080  2014-1000042    2014-10-30                   1   
2359055  2014-1000036    2014-11-13                   1   
2359054  2014-1000036    2014-11-13                   1   
2359053  2014-1000036    2014-11-13                   1   
2359056  2014-1000036    2014-11-13                   1   
2359049  2014-1000033    2014-11-17                   1   
2359048  2014-1000033    2014-11-17                   1   
2359047  2014-1000033    2014-11-17                   1   
2359017  2014-1000014    2014-11-18                   1   
2359018  2014-1000014    2014-11-18                   1   
2359002  2014-1000004    2014-11-05                   1   
2359001  2014-1000004    2014-11-05                   1   

                            nature_mutation  valeur_fonciere  code_postal  \
2358982                               Vente        185000.00      93430.0   
2358981                               Vente        185000.00      93430.0   
2358965                               Vente        400000.00      93380.0   
2358963                               Vente

In [27]:
immo2014 = immo2014.groupby('id_mutation').first().sort_values(by=['id_mutation'], ascending=False)

In [28]:
immo2014.shape

(711594, 18)

In [29]:
immo2014.isna().sum()

date_mutation                     0
numero_disposition                0
nature_mutation                   0
valeur_fonciere                   0
code_postal                    1511
code_commune                      0
nom_commune                       0
code_departement                  0
id_parcelle                       0
code_type_local                   0
type_local                        0
surface_reelle_bati               0
nombre_pieces_principales         0
code_nature_culture          339587
nature_culture               339587
surface_terrain              339588
longitude                     23558
latitude                      23558
dtype: int64

### Group by (  code_commune, nom_commune , type_local, nombre_pieces_principales )

In [32]:
immoGroup2014 = immo2014

In [33]:
immoGroup2014.reset_index( inplace=True)

In [34]:
immoGroup2014.head()

id_mutation date_mutation  numero_disposition  \
0  2014-999999    2014-10-30                   1   
1  2014-999998    2014-11-19                   1   
2  2014-999997    2014-11-05                   1   
3  2014-999996    2014-10-31                   1   
4  2014-999995    2014-11-13                   1   

                      nature_mutation  valeur_fonciere  code_postal  \
0  Vente en l'état futur d'achèvement         242000.0      93240.0   
1                               Vente         116000.0      93300.0   
2                               Vente          29000.0      93300.0   
3                               Vente          64000.0      93400.0   
4                               Vente         110000.0      93300.0   

  code_commune           nom_commune code_departement     id_parcelle  \
0        93072                Stains               93  930720000J0213   
1        93001         Aubervilliers               93  93001000AH0092   
2        93001         Aubervilliers               93  930010000J0059   
3        93070  Saint-Ouen-sur-Seine               93  930700000F0047   
4        93001         Aubervilliers               93  930010000K0177   

   code_type_local   type_local  surface_reelle_bati  \
0              2.0  Appartement                 81.0   
1              2.0  Appartement                 47.0   
2              2.0  Appartement                 20.0   
3              2.0  Appartement                 22.0   
4              2.0  Appartement                 32.0   

   nombre_pieces_principales code_nature_culture nature_culture  \
0                        4.0                 NaN            NaN   
1                        2.0                 NaN            NaN   
2                        2.0                 NaN            NaN   
3                        2.0                 NaN            NaN   
4                        1.0                 NaN            NaN   

   surface_terrain  longitude   latitude  
0              NaN   2.380821  48.958099  
1              NaN   2.392688  48.921551  
2              NaN   2.380943  48.917022  
3              NaN        NaN        NaN  
4              NaN   2.378502  48.912446

In [35]:
cols_to_keep = ["code_commune","nom_commune","code_type_local", "type_local","nombre_pieces_principales","surface_reelle_bati","valeur_fonciere","id_mutation"]

In [36]:
immoGroup2014 = immoGroup2014[cols_to_keep]

In [37]:
immoGroup2014 = immoGroup2014.groupby(["code_commune","nom_commune","type_local","nombre_pieces_principales"]).agg({'surface_reelle_bati':'mean', 'valeur_fonciere':'mean','id_mutation':'size'}).rename(columns={'id_mutation':'nb_mutation'}).reset_index()

In [38]:
immoGroup2014.head()

code_commune              nom_commune type_local  nombre_pieces_principales  \
0        01001  L'Abergement-Clémenciat     Maison                        3.0   
1        01001  L'Abergement-Clémenciat     Maison                        4.0   
2        01001  L'Abergement-Clémenciat     Maison                        5.0   
3        01001  L'Abergement-Clémenciat     Maison                        6.0   
4        01001  L'Abergement-Clémenciat     Maison                        7.0   

   surface_reelle_bati  valeur_fonciere  nb_mutation  
0            62.000000    134750.000000            1  
1           100.000000    110666.666667            3  
2           102.000000    211500.000000            2  
3           119.333333    213273.333333            3  
4           150.000000    272500.000000            1

### Création d'un identifiant de bien

In [ ]:
immoGroup2014['id_bien'] = immoGroup2014["code_commune"] + "_" + immoGroup2014["code_type_local"].astype(int).astype(str) + "_" + immoGroup2014["nombre_pieces_principales"].astype(int).astype(str)

In [ ]:
immoGroup2014.to_csv('../data/immobilier_2014_clean.csv',index=False)